src/Control/Carrier/Resource.hs

```diff
@@ -39,37 +32,31 @@ unliftResource handler = runReader (UnliftIO handler) . runResourceC
 --   . runState @Int 1
 --   $ myComputation
 -- @
-runResource :: MonadUnliftIO m
-            => ResourceC m a
-            -> m a
-runResource r = withRunInIO (\f -> runUnlifting (UnliftIO f) r)
+runResource :: ResourceC m a -> m a
+runResource = runResourceC
 
-newtype ResourceC m a = ResourceC { runResourceC :: ReaderC (UnliftIO m) m a }
+newtype ResourceC m a = ResourceC { runResourceC :: m a }
   deriving (Alternative, Applicative, Functor, Monad, Fail.MonadFail, MonadFix, MonadIO, MonadPlus)
 
-instance MonadUnliftIO m => MonadUnliftIO (ResourceC m) where
-  askUnliftIO = ResourceC . ReaderC $ \(UnliftIO h) ->
-    withUnliftIO $ \u -> pure (UnliftIO $ \r -> unliftIO u (unliftResource h r))
-
 instance MonadTrans ResourceC where
-  lift = ResourceC . lift
+  lift = ResourceC
 
-runUnlifting :: UnliftIO m -> ResourceC m a -> IO a
-runUnlifting h@(UnliftIO handler) = handler . runReader h . runResourceC
```

Comment:
Since we’re using `MonadUnliftIO` directly in the `Carrier`, `runUnlifting`’s role is filled by `unliftIO`, the field selector for `UnliftIO`.